org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java

```diff
@@ -0,0 +1,192 @@
+		}
+	}
+
+	/**
+	 * Provides {@link DocumentSymbol#range range} and {@link DocumentSymbol#selectionRange selection range} for a
+	 * document symbol.
+	 * 
+	 * @see DocumentSymbol#range
+	 * @see DocumentSymbol#selectionRange
+	 */
+	@Beta
+	@Singleton
+	public static class DocumentSymbolRangeProvider {
+		@Inject
+		private DocumentExtensions documentExtensions;
+
+		/**
+		 * The range enclosing this symbol not including leading/trailing whitespace but everything else like comments.
+		 */
+		public Range getRange(EObject object) {
+			Location newFullLocation = documentExtensions.newFullLocation(object);
+			if (newFullLocation != null) {
+				return newFullLocation.getRange();
+			}
+			return null;
+		}
+
+		/**
+		 * The range that should be selected and revealed when this symbol is being picked, e.g the name of a function.
+		 */
+		public Range getSelectionRange(EObject object) {
+			Location newLocation = documentExtensions.newLocation(object);
+			if (newLocation != null) {
+				return newLocation.getRange();
+			}
+			return null;
+		}
+	}
+
+	/**
+	 * Determines whether a document symbol can be marked as {@link DocumentSymbol#deprecated deprecated}.
+	 * 
+	 * @see DocumentSymbol#deprecated
+	 */
+	@Beta
+	@Singleton
+	public static class DocumentSymbolDeprecationInfoProvider {
+		public boolean isDeprecated(EObject object) {
+			return false;
+		}
+
+		public boolean isDeprecated(IEObjectDescription description) {
+			return false;
+		}
+	}
+
+	@Inject
+	private DocumentSymbolMapper.DocumentSymbolNameProvider nameProvider;
+
+	@Inject
+	private DocumentSymbolMapper.DocumentSymbolKindProvider kindProvider;
+
+	@Inject
+	private DocumentSymbolMapper.DocumentSymbolRangeProvider rangeProvider;
+
+	@Inject
+	private DocumentSymbolMapper.DocumentSymbolDetailsProvider detailsProvider;
+
+	@Inject
+	private DocumentSymbolMapper.DocumentSymbolDeprecationInfoProvider deprecationInfoProvider;
+
+	/**
+	 * Converts the {@code EObject} argument into a {@link DocumentSymbol document symbol} without the
+	 * {@link DocumentSymbol#children children} information filled in.
+	 */
+	public DocumentSymbol toDocumentSymbol(EObject object) {
+		DocumentSymbol documentSymbol = new DocumentSymbol();
+		String objectName = nameProvider.getName(object);
+		if (objectName != null) {
+			documentSymbol.setName(objectName);
+		}
+		SymbolKind objectKind = kindProvider.getSymbolKind(object);
+		if (objectKind != null) {
+			documentSymbol.setKind(objectKind);
+		}
+		Range objectRange = rangeProvider.getRange(object);
+		if (objectRange != null) {
+			documentSymbol.setRange(objectRange);
+		}
+		Range objectSelectionRange = rangeProvider.getSelectionRange(object);
+		if (objectSelectionRange != null) {
+			documentSymbol.setSelectionRange(objectSelectionRange);
+		}
+		documentSymbol.setDetail(detailsProvider.getDetails(object));
+		documentSymbol.setDeprecated(deprecationInfoProvider.isDeprecated(object));
+		documentSymbol.setChildren(new ArrayList<>());
```

Comment:
`new ArrayList<>()` -> `Collections.emptyList()`

Reply:
I think elements are added to it later on which will cause exceptions if it's the empty list

Reply:
i did not want to change the semantics here thus kept the list